client.go

```diff
@@ -153,94 +153,181 @@ func readMyCert(configDir string) (string, string, error) {
+		if r.Addr[7:] == "unix://" {
+			raddr, err = net.ResolveUnixAddr("unix", r.Addr[7:])
 		} else {
-			certf, keyf, err := readMyCert(config.ConfigDir)
-			if err != nil {
-				return nil, err
-			}
+			raddr, err = net.ResolveUnixAddr("unix", r.Addr[5:])
+		}
+		if err != nil {
+			return nil, err
+		}
+		return net.DialUnix("unix", nil, raddr)
+	}
+	c.Http.Transport = &http.Transport{Dial: uDial}
+	c.websocketDialer.NetDial = uDial
+	c.Remote = r
 
-			var cert *x509.Certificate
-			if shared.PathExists(c.Config.ServerCertPath(c.Name)) {
-				cert, err = shared.ReadCert(c.Config.ServerCertPath(c.Name))
-				if err != nil {
-					return nil, err
-				}
+	st, err := c.ServerStatus()
+	if err != nil {
+		return err
+	}
+	c.Certificate = st.Environment.Certificate
+	return nil
+}
 
-				c.Certificate = string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
-			}
+func connectViaHttp(c *Client, addr, clientCert, clientKey, serverCert string) error {
 
-			tlsconfig, err := shared.GetTLSConfig(certf, keyf, cert)
-			if err != nil {
-				return nil, err
-			}
+	tlsconfig, err := shared.GetTLSConfigMem(clientCert, clientKey, serverCert)
+	if err != nil {
+		return err
+	}
 
-			tr := &http.Transport{
-				TLSClientConfig: tlsconfig,
-				Dial:            shared.RFC3493Dialer,
-				Proxy:           http.ProxyFromEnvironment,
-			}
+	tr := &http.Transport{
+		TLSClientConfig: tlsconfig,
+		Dial:            shared.RFC3493Dialer,
+		Proxy:           http.ProxyFromEnvironment,
+	}
 
-			c.websocketDialer.NetDial = shared.RFC3493Dialer
-			c.websocketDialer.TLSClientConfig = tlsconfig
+	c.websocketDialer.NetDial = shared.RFC3493Dialer
+	c.websocketDialer.TLSClientConfig = tlsconfig
 
-			if r.Addr[0:8] == "https://" {
-				c.BaseURL = "https://" + r.Addr[8:]
-				c.BaseWSURL = "wss://" + r.Addr[8:]
-			} else {
-				c.BaseURL = "https://" + r.Addr
-				c.BaseWSURL = "wss://" + r.Addr
-			}
-			c.Transport = "https"
-			c.Http.Transport = tr
-			c.Remote = &r
-		}
+	if addr[0:8] == "https://" {
+		c.BaseURL = "https://" + addr[8:]
+		c.BaseWSURL = "wss://" + addr[8:]
 	} else {
-		return nil, fmt.Errorf("unknown remote name: %q", remote)
+		c.BaseURL = "https://" + addr
+		c.BaseWSURL = "wss://" + addr
+	}
+	c.Transport = "https"
+	c.Http.Transport = tr
+	c.Remote = &RemoteConfig{
+		Addr: addr,
+		// TODO: (jam) RemoteConfig.Public is not relevant for the purposes of
+		// an active connection. It only seems to be used to display server
+		// information in "list", though if we aren't getting that information
+		// from the service itself, it seems like it can't be guaranteed to be
+		// correct, as it is just the local information about a remote server.
+		// Is there another reason it is here?
+		Public: false,
+	}
+	// TODO: (jam) It is odd to me that the Unix socket path actually
+	// connects to the service, but the HTTP path just sets up a transport
+	// and socket dialer, but doesn't actually try to connect.
```

Comment:
So I tested this by breaking my 'host.crt' file. And it seems that 'lxc finger' was actually broken in 2.0.0-beta3. In that it doesn't connect at all to the remote host. So I added some changes to finger since we noticed that they are different. I understand Stephane's point that if you don't actually do work on the Client object, then actually doing the SSL handshaking can be a lot of wasted time.